org.eclipse.jdt.ui/core refactoring/org/eclipse/jdt/internal/corext/refactoring/util/ChangedValueChecker.java

```diff
@@ -0,0 +1,578 @@
+import org.eclipse.jdt.core.dom.NullLiteral;
+import org.eclipse.jdt.core.dom.NumberLiteral;
+import org.eclipse.jdt.core.dom.ParenthesizedExpression;
+import org.eclipse.jdt.core.dom.PostfixExpression;
+import org.eclipse.jdt.core.dom.PrefixExpression;
+import org.eclipse.jdt.core.dom.QualifiedName;
+import org.eclipse.jdt.core.dom.SimpleName;
+import org.eclipse.jdt.core.dom.Statement;
+import org.eclipse.jdt.core.dom.StringLiteral;
+import org.eclipse.jdt.core.dom.Type;
+import org.eclipse.jdt.core.dom.WhileStatement;
+
+import org.eclipse.jdt.internal.corext.dom.ASTNodes;
+import org.eclipse.jdt.internal.corext.dom.fragments.IASTFragment;
+import org.eclipse.jdt.internal.corext.util.JavaModelUtil;
+
+public class ChangedValueChecker extends AbstractChecker {
+
+	private ASTNode fNode2;
+
+	private HashSet<Elem> fDependSet;
+
+	private ASTNode fBodyNode;
+
+	private final int THRESHOLD= 2500;
+
+	private ArrayList<ASTNode> fMiddleNodes;
+
+	private boolean fConflict;
+
+	private Set<Position> fPosSet;
+
+	private String fEnclosingMethodSignature;
+
+	public ChangedValueChecker(ASTNode selectedExpression, String enclosingMethodSignature) {
+		this.fEnclosingMethodSignature= enclosingMethodSignature;
+		analyzeSelectedExpression(selectedExpression);
+	}
+
+	public void detectConflict(int startOffset, int endOffset, ASTNode node,
+			ASTNode bodyNode, ArrayList<IASTFragment> candidateList) {
+		fNode2= node;
+		fBodyNode= bodyNode;
+		fConflict= false;
+		fPosSet= Collections.synchronizedSet(new HashSet<>());
+		PathVisitor pathVisitor= new PathVisitor(startOffset, endOffset, fNode2, candidateList);
+		while (fBodyNode != null && (fBodyNode.getStartPosition() + fBodyNode.getLength() < pathVisitor.endOffset
+				|| fBodyNode.getStartPosition() > pathVisitor.startOffset)) {
+			fBodyNode= fBodyNode.getParent();
+		}
+		if (fBodyNode != null)
+			fBodyNode.accept(pathVisitor);
+		fMiddleNodes= pathVisitor.nodes;
+	}
+
+	public void analyzeSelectedExpression(ASTNode selectedExpression) {
+		fDependSet= new HashSet<>();
+		ReadVisitor readVisitor= new ReadVisitor(true);
+		selectedExpression.accept(readVisitor);
+		fDependSet.addAll(readVisitor.readSet);
+	}
+
+	public boolean hasConflict() {
+		ExecutorService threadPool= new ThreadPoolExecutor(5, 10, 5, TimeUnit.SECONDS,
+				new ArrayBlockingQueue<>(10), new ThreadPoolExecutor.CallerRunsPolicy());
+		try {
+			for (ASTNode node : fMiddleNodes) {
+				Position pos= new Position(node.getStartPosition(), node.getLength());
+				if (fPosSet.contains(pos)) {
+					continue;
+				}
+				if (fConflict == true) {
+					break;
+				}
+				threadPool.execute(() -> {
+					fPosSet.add(pos);
+					UpdateVisitor uv= new UpdateVisitor(fDependSet, true);
+					node.accept(uv);
+					if (uv.hasConflict())
+						fConflict= true;
+				});
+			}
+			threadPool.shutdown();
+			threadPool.awaitTermination(5, TimeUnit.SECONDS);
+			while (!threadPool.isTerminated() && fConflict == true) {
```

Comment:
Last comment.  You only want to loop if fConflict is false which means that nothing has found a conflict yet.  Once a conflict is found, there is no need to do any more processing or waiting for termination because nothing more will be found.  That is why I suggested checking if fConflict is true before even asking for termination or shutdown since the finally clause will shutdownNow() regardless.

Reply:
I think I understand what you mean. I modified the code to only call `shutdown` when `fConflict` is false, and the detection would be interrupted if `fConlict` is true.

Reply:
Yes, change looks good.